Condition variable support

hnakamur opened the pull request:
This condition variable implementation is based on the SetEvent solution (section 3.2) at
http://www.cs.wustl.edu/~schmidt/win32-cv-1.html

I also implemented the SignalObjectAndWait solution (section 3.4) at
https://github.com/hnakamur/libuv/commits/cond_vars-SignalObjectAndWait/
Note: this implementation needs that uv_mutex_t implementation to be changed from
CRITICAL_SECTIONs to HANDLEs and APIs such as CreateMutex.
So there is a risk that cause incompatible behavior changes.

Here is a quote from "5. Concluding Remarks" from the above url.

> The SignalObjectsAndWait solution in Section 3.4 is a good approach if fairness is paramount. However, this approach is not as efficient as other solutions, nor is it as portable. Therefore, if efficiency or portability are more important than fairness, the SetEvent approach described in Section 3.2 may be more suitable. Naturally, the easiest solution would be for Microsoft to simply provide condition variables in the Win32 API.

Also my implementation using native CONDITION_VARIABLE is at
https://github.com/hnakamur/libuv/commits/cond_vars/

Preferably, we'd like to use CONDITION_VARIABLE when available, I don't know how to switch implementations, though. Maybe two different binaries for pre-Vista and post-Vista?
Or can we switch two implementations at runtime?

Could you review this pull request? I'd like to get some feedback on this.

hnakamur committed:
Implement condition variable using pthread on Linux.

hnakamur committed:
Implement uv_realtime on darwin. uv_realtime is needed for uv_cond_timedwait.

hnakamur committed:
Implement condition variable using SetEvent. See http://www.cs.wustl.edu/~schmidt/win32-cv-1.html

bnoordhuis commented:
Generally LGTM and something that I've been wanting to add to libuv for a long time now. Nice.

/cc @piscisaureus

hnakamur committed:
Let uv_cond_init() to be consistent with the other _init() functions.

hnakamur committed:
Change pthread_cond_timedwait() return value to 0 for success or -1 on timeout.

hnakamur committed:
Make the condition variable use a monotonic clock on Unixen because the real-time clock is subject to skew.

hnakamur committed:
Add comment on time granularity on Windows and spurious wakeups.

hnakamur commented:
@bnoordhuis thanks for your comments!
I added commits to follow your advices.
Please review them.

bnoordhuis commented:
Good work, nearly there. I'd like for @piscisaureus to review the Windows bits though.

hnakamur committed:
Fix leak of the mutex and event handlers when uv_cond_init() fails.

hnakamur committed:
Abort if WaitForMultipleObjects for an infinite timeout returns WAIT_TIMEOUT.

hnakamur committed:
Fix leak of condvar when uv_cond_init() fails.

hnakamur committed:
Remove lines from note of uv_cond_timedwait, which will only confuse the reader.

hnakamur commented:
@bnoordhuis Thanks again for your comments. I fixed the problems.
@piscisaureus Would you review Windows bits? I'm afraid I'm not an expert on WIndows programming.

piscisaureus commented:
@hnakamur 

> Preferably, we'd like to use CONDITION_VARIABLE when available, I don't know how to switch implementations, though. Maybe two different binaries for pre-Vista and post-Vista?
> Or can we switch two implementations at runtime?

We can, and should, switch between implementations at runtime. We do this in many places already, for example for `SetFileCompletionNotificationModes`. This is done by (trying to) load the function pointer at runtime, and if it is not found (NULL), we use the alternative implementation. You can see how it is done [here](https://github.com/joyent/libuv/blob/7ac23ee0c6ec51974ef7e63679172c567dab14ce/src/win/tcp.c#L113) and [here](https://github.com/joyent/libuv/blob/7ac23ee0c6ec51974ef7e63679172c567dab14ce/src/win/winapi.c#L103-104).

> This condition variable implementation is based on the SetEvent solution
> (section 3.2) at http://www.cs.wustl.edu/~schmidt/win32-cv-1.html
> We could not use the SignalObjectAndWait solution (section 3.4) because
> it want the 2nd argument (type uv_mutex_t) of uv_cond_wait() and
> uv_cond_timedwait() to be HANDLEs, but we use CRITICAL_SECTIONs.

For now I am fine with the solution you propose, if we can use "native" condition variables on Vista+. I don't care much about performance on Windows XP/2k3. If you really want to go down this rabbithole, we could probably re-implement both uv_mutex_t and uv_condvar_t using keyed events (which is the NTAPI facility that underlies critical sections). To get the idea about how this works, see http://locklessinc.com/articles/keyed_events/. It has an example implementation of condition variables.

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +365,146 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
 }
+
+
+/* This condition variable implementation is based on the SetEvent solution
```
See comment https://github.com/joyent/libuv/pull/567#issuecomment-9033115

piscisaureus commented on include/uv-private/uv-win.h:
```diff
@@ -208,6 +208,28 @@ typedef int (WSAAPI* LPFN_WSARECVFROM)
+  unsigned int waiters_count;
+
+  /* Serialize access to <waiters_count>. */
+  HANDLE waiters_count_lock;
```
For performance reasons I would much rather use a critical section for this.

hnakamur replied:
Me too. And the original implementation at  http://www.cs.wustl.edu/~schmidt/win32-cv-1.html does use a critical section.
However, I got an error like "`condvar_1` failed: exit code -1073741819"
https://github.com/hnakamur/libuv/commit/e8beca68aff94c1acb27fb97d0265fa9d5235cd2

I will try again using a critical section.

piscisaureus replied:
Well, that should be a writing on the wall. The contents of this field probably get corrupted somewhere.

hnakamur replied:
I tried again and this time it works with CRITICAL_SECTION.

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +365,146 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
+    abort();
+  if (!CloseHandle(cond->waiters_count_lock)) {
+    /* Note: I don't know why but I always get ERROR_INVALID_HANDLE here.
+     * I neglect it.
```
I don't really think that is acceptable... I can dig down but I would suggest to use a critical section for `waiters_count_lock`.

hnakamur replied:
Thanks. I switched to a critical section.

piscisaureus commented on include/uv-private/uv-win.h:
```diff
@@ -208,6 +208,28 @@ typedef int (WSAAPI* LPFN_WSARECVFROM)
+  HANDLE waiters_count_lock;
+
+  /* Signal and broadcast event HANDLEs. */
+  HANDLE events[UV_COND_MAX_EVENTS];
```
I can see why you wanted to put this in an array, but this is really obtuse and not really worth it, since you would only need to copy 2 HANDLEs when you had just used named struct members.

hnakamur replied:
I think it is needed because we would like to use WaitForMultipleObjects
https://github.com/hnakamur/libuv/blob/1c4a63028438539c452f13d1729b63f64e9a519b/src/win/thread.c#L471

Or is it safe to use WaitForMultipleObjects for 2 HANDLEs in struct?

piscisaureus replied:
@hnakamur It's probably safe but pretty hacky too. I suppose you can just do:

``` c
HANDLE handles[2] = { bla->foo_event, bla->bar_event };
WaitForMultipleObject(&handles, 2, ...);
```

hnakamur replied:
Thanks for your proposal. I fix the implementation at https://github.com/hnakamur/libuv/commit/8b698b8ea851e3813ddf49069942c4f0e06c43cb

hnakamur commented:
@piscisaureus Thanks for your comments. I don't care much about performance on Windows XP/2k3 too. I would definitely like switching implementations and I'm going to try that.

hnakamur committed:
Switch waiters_count_lock to CRITICAL_SECTION.

hnakamur committed:
Use native condition variables on Vista+ or use fallback on pre-Vista.

hnakamur committed:
Have two distinct struct members for signal and broadcast event.

hnakamur committed:
Use the same style as rwlock in switch implementations of condition variable.

hnakamur committed:
Call uv_loop_new in tests for uv_winapi_init.

hnakamur committed:
Fix bugs in implemnentation using native condition variable.

hnakamur commented:
@piscisaureus Thanks for your comment. I think now the code is in good shape. Could you review it?

/cc @bnoordhuis

bnoordhuis commented on src/unix/darwin.c:
```diff
@@ -41,6 +41,9 @@
 #include <unistd.h>  /* sysconf */
 
+#undef NANOSEC
+#define NANOSEC ((uint64_t) 1e9)
```
Leftover?

bnoordhuis commented on test/test-condvar.c:
```diff
@@ -0,0 +1,183 @@
+  uv_thread_t thread;
+  worker_config wc;
+
+  uv_loop_new(); /* for uv_winapi_init */
```
Can this be worked around, e.g. by calling `uv__once_init()` in `uv_cond_init()`?

bnoordhuis replied:
@piscisaureus I think that applies to `uv_rwlock_init()` and friends as well. They work by accident now.

piscisaureus replied:
Agreed

bnoordhuis commented:
LGTM. The calls to `uv_loop_new()` in the tests make valgrind complain about memory leaks but that's easily fixed by replacing it with a call to `uv_default_loop()` or calling `uv_loop_delete()` at the end of the test.

bnoordhuis commented:
One last comment, can you apply the patch below?

``` diff
diff --git a/uv.gyp b/uv.gyp
index c9c6a58..622b234 100644
--- a/uv.gyp
+++ b/uv.gyp
@@ -305,6 +305,7 @@
         'test/test-mutexes.c',
         'test/test-signal.c',
         'test/test-thread.c',
+        'test/test-condvar.c',
         'test/test-timer-again.c',
         'test/test-timer.c',
         'test/test-tty.c',
```

piscisaureus commented on src/win/winapi.h:
```diff
@@ -4426,6 +4426,25 @@ typedef VOID (WINAPI* sReleaseSRWLockShared)
+typedef VOID (WINAPI* sInitializeConditionVariable)
+             (PCONDITION_VARIABLE ConditionVariable);
+
+typedef WINBOOL (WINAPI* sSleepConditionVariableCS)
```
this should be `BOOL`  and not `WINBOOL`

piscisaureus commented on src/win/winapi.h:
```diff
@@ -4426,6 +4426,25 @@ typedef VOID (WINAPI* sReleaseSRWLockShared)
+                 PCRITICAL_SECTION CriticalSection,
+                 DWORD dwMilliseconds);
+
+typedef WINBOOL (WINAPI* sSleepConditionVariableSRW)
```
ditto

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +383,215 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
+
+void uv_cond_destroy(uv_cond_t* cond) {
+  if (HAVE_CONDVAR_API())
+    return uv_cond_condvar_destroy(cond);
```
Void function can't return a value (atleast not in msvc)

bnoordhuis replied:
Sure it can!

``` c
return uv_cond_condvar_destroy(cond), (void) 0;
```

/me ducks

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +383,215 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
+  if (HAVE_CONDVAR_API())
+    return uv_cond_condvar_destroy(cond);
+  else
+    return uv_cond_fallback_destroy(cond);
```
ditto

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +383,215 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
+  /* Wait for either event to become signaled due to <pthread_cond_signal>
+   * being called or <pthread_cond_broadcast> being called.
+   */
+  HANDLE handles[2] = {
```
c89: variable declarations must be at the top of a scope.

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +383,215 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
+      FALSE, /* non-signaled initially */
+      NULL); /* unnamed */
+  if (!cond->fallback.signal_event)
+    return -1;
```
nit: DeleteCriticalSection() the critical section when it fails?

piscisaureus commented on src/win/thread.c:
```diff
@@ -365,3 +383,215 @@ inline static int uv__rwlock_fallback_trywrlock(uv_rwlock_t* rwlock) {
+      FALSE, /* non-signaled initially */
+      NULL); /* unnamed */
+  if (!cond->fallback.broadcast_event)
+    goto error;
```
ditto

piscisaureus commented:
Mostly good. See comments.

hnakamur committed:
Remove the leftover code.

hnakamur committed:
Add test/test-condvar.c to the gyp build system.

hnakamur committed:
Use uv__once_init() instead of uv_loop_new()

hnakamur committed:
Use BOOL instead of WINBOOL.

hnakamur committed:
Fix leak in cleanup for failure in _init(). Remove return for void functions. Move variable declarations to the top of a scope.

hnakamur committed:
Don't use pthread_condattr_setclock() in OSX because it does not exist.
Here is quote from man pthread_condattr_init.
DESCRIPTION
     Condition attribute objects are used to specify parameters to
     pthread_cond_init().  FreeBSD's implementation of conditions does not
     support any non-default attributes, so these functions are not very use-
     ful, though they are required to to be present by POSIX.
Note: The condvar_4 test fails on OSX now.

hnakamur committed:
Use gettimeofday instead of hrtime on OSX, following an example in man pthread_cond_timeout.
Add comment on the precision of timeout on OSX.

hnakamur commented:
@bnoordhuis @piscisaureus Thanks for your comments. I have added more commits.
Also I noticed OSX does not have pthread_condattr_setclock() and condvar_4 test fails if using hrtime in uv_cond_timedwait. I switched gettimeofday instead of hrtime following an example in man pthread_cond_timedwait.

piscisaureus commented:
almost... almost...

The uv__once_init patch is not what I meant. Instead you should just make uv_condvar_init call that function (see 744dc3e77).

Also, the return from a void function issue is still there.

I need @bnoordhuis to comment on the hrtime issue.

piscisaureus commented:
> Also, the return from a void function issue is still there.

Sorry, I lied. You solved it.

hnakamur committed:
Call uv__once_init in uv_cond_init instead of test codes.

hnakamur committed:
Fix document on timeout of uv_cond_timedwait

hnakamur committed:
Add a blank line before #if

hnakamur committed:
Add comment in uv.h on timeout resolution of uv_cond_timedwait on OSX.

hnakamur commented:
@bnoordhuis @piscisaureus Thanks for your comments again. I added more commits.
Could you review them? And I have to go now. I'll check your review when I'm back home.

piscisaureus commented:
@hnakamur 

Ok, it builds now for me (I use msvc). I have 2 more requests:

(1)

I get two warnings while building:

```
test\test-condvar.c(114): warning C4244: 'function' : conversion from 'double' to 'uint64_t', possible loss of data [D:\libuv\run-tests.vcxproj]
test\test-condvar.c(139): warning C4244: 'function' : conversion from 'double' to 'uint64_t', possible loss of data [D:\libuv\run-tests.vcxproj]
```

Although these are harmless, it would be nice to add an explicit cast to avoid them.

(2)

I like to keep the number of "windows" headers in uv-win as low as possible. I know there are quite a few already, but these were really unavoidable. Other headers can go into winapi.h (if you really need them; I don't think you do here).

I'd like to see most of the stuff you added there go. I think doing just this would suffice:

``` c
#ifndef RTL_CONDITION_VARIABLE_INIT
  typedef PVOID RTL_CONDITION_VARIABLE;
#endif
```

Please put it together with the other windows headers, so maybe right below `#ifndef _NTDEF_` ?

hnakamur committed:
Add explicit casts from double to uint64_t for timeout.

hnakamur committed:
Cleanup definitions in uv-win.h

hnakamur commented:
@piscisaureus Thanks again for your comments.
As for (1), I agree with you and fixed them.
As for (2), I am not sure what you exactly mean. So I took a guess to modify uv-win.h.
Could you review and verify it?

piscisaureus commented:
@hnakamur I added what I had in mind in #575. I will wait for @bnoordhuis to give a final thumbs up on the unix bits (he's at a conference today). Thanks a lot for your work!

hnakamur commented:
@piscisaureus Thanks for your reviews and modification in #575!
@bnoordhuis I'm looking forward to see your comments!

piscisaureus commented:
Landed in 976c8a4
